#summary PSI-MI TAB 2.7 format reference

= PSI-MI TAB 2.7 Format =

== Introduction ==

The MITAB27 format is an extention of the PSI-MI 2.5 (1) and 2.6 standard. It has been derived from the tabular format provided by !BioGrid. MITAB27 only describes binary interactions, one pair of interactors per row. Columns are separated by tabulations. Tools allowing to manipulate this data format are available (2). 

(1) http://www.pubmedcentral.nih.gov/articlerender.fcgi?tool=pubmed&pubmedid=17925023

(2) http://www.psidev.info/index.php?q=node/60#tools


== Column definitions ==

The column contents should be as follows:

  #  *Unique identifier for interactor A*, represented as databaseName:ac, where     databaseName is the name of the corresponding database as defined in the [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI:0444&termName=database%20citation PSI-MI controlled vocabulary], and ac is the unique primary identifier of the molecule in the database. Even though identifiers from multiple databases can be separated by "|", it is recommended to give only one identifier in this column. It is recommended that proteins be identified by stable identifiers such as their UniProtKB or !RefSeq accession number. Small molecules should have Chebi identifiers, nucleic acids should have embl/ddbj/genbank identifiers and gene should have entrez gene/locuslink, ensembl, or ensemblGenome identifiers. This column should never be empty except for describing intra-molecular interactions or auto-catalysis.
  #  *Unique identifier for interactor B*.
  #  *Alternative identifier for interactor A*, represented as databaseName:ac, where     databaseName is the name of the corresponding database as defined in the [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI:0444&termName=database%20citation PSI-MI controlled vocabulary], and ac is the primary identifier of the molecule in the database. Multiple identifiers separated by "|". It is recommended to only give database identifiers in this column. Other cross references for interactor A such as GO xrefs should be moved to the column 'Interactor xrefs A' and interactor names such as gene names should be moved to the column 'Alias A'
  # *Alternative identifier for interactor B*.
  #  *Aliases for A*, separated by "|". Representation as databaseName:name(alias type), where     databaseName is the name of the corresponding database as defined in the [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI:0444&termName=database%20citation PSI-MI controlled vocabulary], name is the alias name and alias type is the name of the corresponding alias type as defined in the [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A0300&termName=alias%20type PSI-MI controlled vocabulary]. In the absence of databaseName, one can use `unknown`. Multiple names separated by "|". In parenthesis, 'display_short' and 'display_long' are used to describe what name can be used for network display.
  #  *Aliases for B*.
  #  *Interaction detection methods*, taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI:0001&termName=interaction%20detection%20method PSI-MI controlled Vocabulary], and represented as databaseName:identifier(methodName), separated by "|". As the detection methods are taken from the PSI-MI ontology, the database name is 'psi-mi'. Interaction detection method is recommended by MIMIx and can be used for scoring interactions so it is recommended to always give this information. It is also recommended to give one interaction detection method per MITAB line (for clustering purposes).
  #  *First author* surname(s) of the publication(s) followed by 'et al.' and the publication year in parenthesis, e.g. "Ciferri et al.(2005)". Separated by "|".
  #  *Identifier of the publication* in which this interaction has been shown. Database name taken from the [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI:0445&termName=literature%20database PSI-MI controlled vocabulary], represented as databaseName:identifier. Multiple identifiers separated by "|". The publication identifier is used for clustering interactions from different data providers in PSICQUIC and can be used for scoring interactions so it is recommended to always give this information. It is recommended to give one pubmed id per MITAB line and IMEx ids can be added.
  # *NCBI Taxonomy identifier for interactor A*. represented as taxid:identifier(organism name) where the identifier is the taxon id of the organism and organism name can either be the common name or scientific name. Even though multiple identifiers can be separated by "|", it is recommended to have one organism per interactor per MITAB line. If both scientific name and common name are given, they should be represented with : "taxid:id1(common name1)|taxid:id1(scientific name1)". Note: Currently no taxonomy identifiers other than NCBI taxid are anticipated, apart from the use of -1 to indicate "in vitro", -2 to indicate "chemical synthesis", -3 indicates "unknown", -4 indicates "in vivo" and -5 indicates "in silico". It is recommended to always give this information for proteins and genes. For small molecules and nucleic acids, this information should be given when available.
  # *NCBI Taxonomy identifier for interactor B*.
  # *Interaction types*, taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI:0190&termName=interaction%20type PSI-MI controlled vocabulary], and represented as dataBaseName:identifier(interactionType), separated by "|". As the detection methods are taken from the PSI-MI ontology, the database name is 'psi-mi'. Interaction type can be used for scoring interactions so it is recommended to always give this information. It is also recommended to give one interaction type per MITAB line (for clustering purposes).
  # *Source databases* and identifiers, taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI:0444&termName=database%20citation PSI-MI controlled vocabulary], and represented as databaseName:identifier(sourceName). As the detection methods are taken from the PSI-MI ontology, the database name is 'psi-mi'. Multiple source databases can be separated by "|". When the interaction has been imported and reported by different sources, it is recommended to give the original source plus the source that currently reports the interaction.
  # *Interaction identifier(s)* in the corresponding source database, represented by databaseName:identifier. It is recommended to always give a unique identifier per interaction (binary and n-ary) so in case of complexes which have been expanded, it would be possible to retrieve and re-build the original complex based on the interaction identifier.
  # *Confidence score*. Denoted as scoreType:value where scoreType is taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A1064&termName=interaction%20confidence PSI-MI controlled vocabulary]. Multiple scores separated by "|".
  # *Expansion*. Model used to convert n-ary interactions into binary interactions for purpose of export in MITAB file. The known expansions are none (if true binary interaction), spoke, matrix or bipartite.
  # *Biological role A* , taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A0500&termName=biological%20role PSI-MI controlled vocabulary], and represented as dataBaseName:identifier(biological role name), separated by "|".
  # *Biological role B* , taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A0500&termName=biological%20role PSI-MI controlled vocabulary], and represented as dataBaseName:identifier(biological role name), separated by "|".
  # *Experimental role A* , taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A0495&termName=experimental%20role PSI-MI controlled vocabulary], and represented as dataBaseName:identifier(experimental role name), separated by "|".
  # *Experimental role B* , taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A0495&termName=experimental%20role PSI-MI controlled vocabulary], and represented as dataBaseName:identifier(experimental role name), separated by "|".
  # *Interactor type A* , taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A0313&termName=interactor%20type PSI-MI controlled vocabulary], and represented as dataBaseName:identifier(interactor type name), separated by "|".
  # *Interactor type B* , taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A0313&termName=interactor%20type PSI-MI controlled vocabulary], and represented as dataBaseName:identifier(interactor type name), separated by "|".
  #  *Xref for interactor A*, for example the gene ontology cross references associated. Representation as databaseName:identifier(text). Multiple cross references separated by "|".
  #  *Xref for interactor B*, for example the gene ontology cross references associated. Representation as databaseName:identifier(text). Multiple cross references separated by "|".
  #  *Xref for the interaction*, for example the gene ontology cross references associated (components, etc.) or OMIM cross references. Representation as databaseName:identifier(text). Multiple cross references separated by "|".
  #  *Annotations for interactor A*, for example comments about this interactor. Representation as topic:text. Multiple annotations separated by "|".
  #  *Annotations for Interactor B*, for example comments about this interactor. Representation as databaseName:identifier(text). Multiple annotations separated by "|".
  #  *Annotations for the interaction*, for example comments about this interaction. Representation as topic:text. Multiple annotations separated by "|".
  # *NCBI Taxonomy identifier for the host organism*. Database name for NCBI taxid taken from the [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI:0444&termName=database%20citation PSI-MI controlled vocabulary], represented as databaseName:identifier (typicaly databaseName is set to 'taxid'). Multiple identifiers separated by "|". Note: In this column, the databaseName:identifier(speciesName) notation is only there for consistency. Currently no taxonomy identifiers other than NCBI taxid are anticipated, apart from the use of -1 to indicate "in vitro", -2 to indicate "chemical synthesis", -3 indicates "unknown", -4 indicates "in vivo" and -5 indicates "in silico".
  #  *Parameters of the interaction*, for example kinetics. Representation as type:value(text). The type can be taken from the corresponding [http://www.ebi.ac.uk/ontology-lookup/browse.do?ontName=MI&termId=MI%3A0640&termName=parameter%20type PSI-MI controlled vocabulary]. Multiple parameters separated by "|".
  #  *Creation date* Representation as yyyy/mm/dd.
  #  *Update date* Representation as yyyy/mm/dd.
  #  *Checksum for interactor A*, for instance the ROGID of the interactor. Representation as databaseName:checksum
  #  *Checksum for interactor B*, for instance the ROGID of the interactor. Representation as databaseName:checksum
  #  *Checksum for interaction*, for instance the RIGID of the interaction. Representation as databaseName:checksum
  #  *negative* Boolean value to distinguish positive interactions (negative = false) from negative interactions (negative = true)
  # *Binding domain for interactor A*: feature_type:range(text), for instance sufficient to bind:27-195,201-133 (IPR000785). The use of the following characters is allowed to describe a range: ‘?’, ‘n’, ‘>1’ etc.
  # *Binding domain for interactor B*: feature_type:range(text), for instance sufficient to bind:27-195,201-133 (IPR000785). The use of the following characters is allowed to describe a range: ‘?’, ‘n’, ‘>1’ etc.
  # *Stoichiometry for interactor A*: A numerical value describing the count of instance of the molecule participanting in the interaction. Example: for self interactors e.g. a kinase occluding its kinase domain by an internal phospho-tyrosine/SH2 domain interaction, both Interactor A/B columns will show the same molecule accession number with the stoichiometry 1 and 0.
  # *Stoichiometry for interactor B*: A numerical value describing the count of instance of the molecule participanting in the interaction. Example: for self interactors e.g. a kinase occluding its kinase domain by an internal phospho-tyrosine/SH2 domain interaction, both Interactor A/B columns will show the same molecule accession number with the stoichiometry 1 and 0.
  # *Participant identification method for interactor A*: 
  # *Participant identification method for interactor B*: 

All the columns are mandatory.

== Syntax ==

Columns are normally formed by fields delimited by "|", with a structure like this one:

{{{
<XREF>:<VALUE>(<DESCRIPTION>)
}}}

Due to the unsafe use of reserved characters in the values, we have recently added the possibility to surround `<XREF>`, `<VALUE>` or `<DESCRIPTION>` with quotes if they contain a special symbol.

In MI-TAB, the reserved characters are: 

{{{
|
(
)
:
\t (tabulation)
}}}

Whenever this happen in your data, surround the value with double quotes:

{{{
"<XREF_WITH_RESERVED_CHARS>":"<VALUE_WITH_RESERVED_CHARS>"("<DESCRIPTION>")
}}}

Note that the quotes are before and after each part. The escaped data should look like in the following examples:

{{{
psi-mi:"MI:0000"(a cv term)
psi-mi:"MI:0000"("I can now use braces ()()() or pipes ||| here and ::colons::")
}}}
If you want to use a quote within a quote, escape it:

{{{
uniprotkb:P12345("a \"nice\" protein")
}}}